### PR TITLE
[pickers] Migrate `AdapterDayjs` to our repository

### PR DESCRIPTION
**Pull request by flaviendelangle**
- Migrate all the run on the `dayjs` adapter to our repo
- Migrate all the methods of `AdapterDayjs` to our repo
- Remove the dependency on `@date-io/dayjs`

### CODE REVIEW

File: packages/x-date-pickers/src/models/adapters.ts [hunk 1]
```diff
@@ -153,7 +153,7 @@ export interface MuiPickersAdapter<TDate> {
    * @param {string} format The format to expand.
```
Comment (flaviendelangle) [inline packages/x-date-pickers/src/models/adapters.ts:153]: I just improved the order of the methods and deprecated `isNull` which is not very usefull.

**Comment by alexfauquette** [comment]
Do you think it would be possible to ensure we have a 100% test coverage on the adapter code? maybe a `codecov.yml` file could do the job

**Comment by flaviendelangle** [comment]
I can migrate the code coverage infra from `date-io`
Would you agree to not add tests for the methods not covered (the one that were already on our repo) in this PR ?
I agree that we should improve the code coverage after, especially with timezones.

**Comment by alexfauquette** [comment]
> Would you agree to not add tests for the methods not covered (the one that were already on our repo) in this PR ?

Yes, my main concern was about we don't have incentive to improve testing, and so we say in the state where "Yes, it would be great but latter"

File: packages/x-date-pickers/src/AdapterDayjs/AdapterDayjs.test.tsx [hunk 2]
```diff
@@ -37,7 +42,79 @@ const localizedTexts = {
...
+        expect(result[0][0].format('dd')).to.equal('Su');
+      });
+
+      it('is12HourCycleInCurrentLocale: properly determine should use meridiem or not', () => {
+        expect(adapter.is12HourCycleInCurrentLocale()).to.equal(true);
+      });
+    });
+
+    it('Formatting', () => {
+      const adapter = new AdapterDayjs({ locale: 'en-US' });
```
Comment (LukasTy) [inline packages/x-date-pickers/src/AdapterDayjs/AdapterDayjs.test.tsx:91]: Nitpick, on L72 we use `locale: 'en'` and here it's `locale: 'en-US'`. Would it make sense sticking to one or the other? 🤔

**Review by LukasTy — APPROVED** [review]
Great work! 🙌

**Comment by flaviendelangle** [comment]
@alexfauquette I added a 95% code coverage threshold to the adapter.
We can work toward having 100% (and also add better coverage to other parts of the codebase, even though I don't think we should target code coverage everywhere).

**Review by alexfauquette — APPROVED** [review]

**Pull request merged**